R/commonQualityControl.R

```diff
@@ -444,10 +453,190 @@ NelsonLaws <- function(data, allsix = FALSE, chart = "i", xLabels = NULL) {
   return(table)
 }
 
-decimalplaces <- function(x) {
-  if ((x %% 1) != 0) {
-    nchar(strsplit(sub('0+$', '', as.character(x)), ".", fixed=TRUE)[[1]][[2]])
+.decimalplaces <- function(x) {
+  x <- unlist(x)
+  nDecimals <- numeric(length(x))
+  for(i in seq_along(x)) {
+    if (round(x[i] %% 1, 10) != 0) {   # never more than 10 decimals
+      formattedx <- format(x[i], scientific = FALSE)
+      nDecimals[i] <- nchar(strsplit(sub('0+$', '', as.character(formattedx)), ".", fixed=TRUE)[[1]][[2]])
+    } else {
+      nDecimals[i] <- 0
+    }
+  }
+  return(nDecimals)
+}
+
+.IMRchart <- function(dataset, options, variable = "", measurements = "", cowPlot = FALSE, manualXaxis = "", Wide = FALSE,
+                      stages = "") {
+
+  if (identical(stages, "")) {
+    nStages <- 1
   } else {
-    return(0)
+    nStages <- length(unique(dataset[[stages]]))
   }
+  
+  ppPlot <- createJaspPlot(width = 900 + nStages * 100, height = 650)
+  
+  # Calculate values
+  dataPlotI <- list(allValues = list())
+  dataPlotR <- list(allValues = list())
+  for (i in seq_len(nStages)) {
+    if (identical(stages, "")) {
+      dataForPlot <- dataset
+    } else {
+      dataForPlot <- subset(dataset, dataset[[stages]] == unique(dataset[[stages]])[i])
+    }
+    
+    if (identical(measurements, "") && !identical(variable, "")) {
+      ppPlot$dependOn(optionContainsValue = list(variables = variable))
+      data <- data.frame(process = dataForPlot[[variable]])
+      sixsigma_I <- qcc::qcc(data$process, type ='xbar.one', plot=FALSE)
+      xmr.raw.r <- matrix(cbind(data$process[1:length(data$process)-1], data$process[2:length(data$process)]), ncol = options$ncol)
```

Comment:
I don't see the difference.

```
aaa <- 1:10; 
aaa[1:length(aaa)-1] == aaa[1:(length(aaa)-1)]
```

Still already thanks for pointing this out. I think the idea here was to create a matrix that per row contains the current observation and the next n (ncol = moving range length) observations, so, if ncol = 2: 

```
      [,1] [,2]
 [1,]    1    2
 [2,]    2    3
 [3,]    3    4
 [4,]    4    5
```

However, with anything else than ncol = 2 this goes completly wrong. E.g. with ncol = 3 we get:

```
 [,1] [,2] [,3]
[1,]    1    7    5
[2,]    2    8    6
[3,]    3    9    7
[4,]    4    2    8
[5,]    5    3    9
[6,]    6    4   10
```

instead of

```
 [,1] [,2] [,3]
[1,]    1    2    3
[2,]    2    3    4
[3,]    3    4    5
...
```
 Also confirmed with Minitab that the results only match for ncol = 2. So this part will need a complete rewrite.

Reply:
Okay, I get your point now. It's a little weird that this works:

```
aaa <- 1:10; 
aaa[1:9] == aaa[0:9]
```